File: bindings/python/tests/test_geometry_frame.py
"""Tests for GeometryFrame public API: decode_frame, encode_frame, to_geometry, to_buffer, from_buffer."""

import numpy as np
import pytest
import shapely
import wkp
from shapely.geometry import LineString, MultiPolygon, Point, Polygon


# ---------------------------------------------------------------------------
# decode_frame / encode_frame roundtrip
# ---------------------------------------------------------------------------

@pytest.mark.parametrize(
    "geom",
    [
        Point(174.776, -41.289),
        LineString([(0, 0), (1, 1), (2, 2)]),
        Polygon([(0, 0), (10, 0), (10, 10), (0, 10), (0, 0)], [[(2, 2), (4, 2), (4, 4), (2, 4), (2, 2)]]),
        shapely.multipoints([(0, 0), (1, 1), (2, 2)]),
        shapely.multilinestrings([[(0, 0), (1, 1)], [(2, 2), (3, 3)]]),
        MultiPolygon([
            Polygon([(0, 0), (2, 0), (2, 2), (0, 2), (0, 0)]),
            Polygon([(4, 4), (8, 4), (8, 8), (4, 8), (4, 4)]),
        ]),
    ],
)
def test_decode_frame_roundtrip(geom):
    encoded = wkp.encode(geom, precision=6)
    frame = wkp.decode_frame(encoded)

    assert frame.version == 1
    assert frame.precision == 6
    assert frame.dimensions == 2
    assert isinstance(frame.coords, np.ndarray)
    assert frame.coords.dtype == np.float64
    assert frame.coords.ndim == 2
    assert frame.coords.shape[1] == 2
    assert isinstance(frame.segment_point_counts, np.ndarray)
    assert isinstance(frame.group_segment_counts, np.ndarray)

    # Re-encode from frame and compare
    re_encoded = wkp.encode_frame(frame)
    assert re_encoded == encoded


def test_decode_frame_explicit_context():
    ctx = wkp.Context()
    geom = LineString([(0, 0), (1, 1)])
    encoded = wkp.encode(geom, precision=5, ctx=ctx)
    frame = wkp.decode_frame(encoded, ctx=ctx)
    assert frame.precision == 5


def test_decode_frame_coords_shape():
    # Polygon with shell + hole: 5 + 5 = 10 points total
    geom = Polygon([(0, 0), (10, 0), (10, 10), (0, 10), (0, 0)], [[(2, 2), (4, 2), (4, 4), (2, 4), (2, 2)]])
    encoded = wkp.encode(geom, precision=6)
    frame = wkp.decode_frame(encoded)

    assert frame.coords.shape == (10, 2)
    assert len(frame.segment_point_counts) == 2   # 2 rings
    assert len(frame.group_segment_counts) == 1   # 1 polygon group
    assert sum(frame.segment_point_counts) == 10


# ---------------------------------------------------------------------------
# to_geometry
# ---------------------------------------------------------------------------

def test_frame_to_geometry_roundtrip():
    geom = Polygon([(0, 0), (10, 0), (10, 10), (0, 10), (0, 0)], [[(2, 2), (4, 2), (4, 4), (2, 4), (2, 2)]])
    encoded = wkp.encode(geom, precision=6)
    frame = wkp.decode_frame(encoded)
    recovered = frame.to_geometry()
    assert recovered.equals_exact(geom, tolerance=1e-6)


# ---------------------------------------------------------------------------
# to_buffer / from_buffer
# ---------------------------------------------------------------------------

def test_buffer_roundtrip_linestring():
    geom = LineString([(1.1, 2.2), (3.3, 4.4), (5.5, 6.6)])
    encoded = wkp.encode(geom, precision=6)
    frame = wkp.decode_frame(encoded)

    buf = frame.to_buffer()
    assert isinstance(buf, bytes)

    recovered = wkp.GeometryFrame.from_buffer(buf)
    assert recovered.version == frame.version
    assert recovered.precision == frame.precision
    assert recovered.dimensions == frame.dimensions
    assert recovered.geometry_type == frame.geometry_type
    np.testing.assert_array_equal(recovered.coords, frame.coords)
    np.testing.assert_array_equal(recovered.segment_point_counts, frame.segment_point_counts.astype(np.uint64))
    np.testing.assert_array_equal(recovered.group_segment_counts, frame.group_segment_counts.astype(np.uint64))


def test_buffer_roundtrip_multipolygon():
    geom = MultiPolygon([
        Polygon([(0, 0), (2, 0), (2, 2), (0, 2), (0, 0)]),
        Polygon([(4, 4), (8, 4), (8, 8), (4, 8), (4, 4)]),
    ])
    encoded = wkp.encode(geom, precision=5)
    frame = wkp.decode_frame(encoded)

    buf = frame.to_buffer()
    recovered = wkp.GeometryFrame.from_buffer(buf)

    # Verify geometry can be recovered
    geom_recovered = recovered.to_geometry()
    assert geom_recovered.equals_exact(geom, tolerance=1e-5)


def test_buffer_header_size():
    geom = Point(1.0, 2.0)
    encoded = wkp.encode(geom, precision=6)
    frame = wkp.decode_frame(encoded)
    buf = frame.to_buffer()

    # Header is 40 bytes: 4×int32 + 3×uint64 = 16 + 24 = 40
    assert len(buf) >= 40
    # Total = 40 + coord_value_count*8 + (segment_count + group_count)*4
    ncoords = frame.coords.size
    nseg = len(frame.segment_point_counts)
    ngrp = len(frame.group_segment_counts)
    expected_size = 40 + ncoords * 8 + nseg * 4 + ngrp * 4
    assert len(buf) == expected_size


def test_buffer_from_bytes_and_memoryview():
    geom = LineString([(0, 0), (1, 1)])
    encoded = wkp.encode(geom, precision=4)
    frame = wkp.decode_frame(encoded)
    buf = frame.to_buffer()

    # Works from bytes
    r1 = wkp.GeometryFrame.from_buffer(buf)
    # Works from memoryview
    r2 = wkp.GeometryFrame.from_buffer(memoryview(buf))
    np.testing.assert_array_equal(r1.coords, r2.coords)

File: bindings/javascript/packages/web/tests/test-geometry-frame.mjs
import test from 'node:test';
import assert from 'node:assert/strict';

import { createWkp } from '../src/index.js';

function assertClose(a, b, epsilon = 1e-6) {
    assert.ok(Math.abs(a - b) <= epsilon, `expected ${a} ≈ ${b} (epsilon ${epsilon})`);
}

function assertGeometryClose(actual, expected, epsilon = 1e-6) {
    assert.equal(actual.type, expected.type);
    function walk(a, e) {
        if (Array.isArray(e)) {
            assert.ok(Array.isArray(a));
            assert.equal(a.length, e.length);
            for (let i = 0; i < e.length; i += 1) walk(a[i], e[i]);
            return;
        }
        assertClose(a, e, epsilon);
    }
    walk(actual.coordinates, expected.coordinates);
}

const GEOMETRIES = [
    { type: 'Point', coordinates: [174.776, -41.289] },
    { type: 'LineString', coordinates: [[0, 0], [1, 1], [2, 2]] },
    { type: 'Polygon', coordinates: [[[0, 0], [1, 0], [1, 1], [0, 1], [0, 0]]] },
    { type: 'MultiPoint', coordinates: [[0, 0], [1, 1]] },
    { type: 'MultiLineString', coordinates: [[[0, 0], [1, 1]], [[2, 2], [3, 3]]] },
    {
        type: 'MultiPolygon',
        coordinates: [
            [[[0, 0], [1, 0], [1, 1], [0, 1], [0, 0]]],
            [[[2, 2], [3, 2], [3, 3], [2, 3], [2, 2]]]
        ]
    },
];

test('decodeFrame returns GeometryFrame for all geometry types (web)', async () => {
    const wkp = await createWkp();

    for (const geometry of GEOMETRIES) {
        const encoded = wkp.encode(geometry, 6);
        const frame = wkp.decodeFrame(encoded);

        assert.ok(frame instanceof wkp.GeometryFrame, `expected GeometryFrame for ${geometry.type}`);
        assert.equal(frame.version, 1);
        assert.equal(frame.precision, 6);
        assert.equal(frame.dimensions, 2);
        assert.ok(frame.coords instanceof Float64Array);
        assert.ok(frame.segmentPointCounts instanceof Uint32Array);
        assert.ok(frame.groupSegmentCounts instanceof Uint32Array);
    }
});

test('encodeFrame roundtrip produces same output as encode (web)', async () => {
    const wkp = await createWkp();

    for (const geometry of GEOMETRIES) {
        const encoded = wkp.encode(geometry, 6);
        const frame = wkp.decodeFrame(encoded);
        const reEncoded = wkp.encodeFrame(frame);

        assert.equal(reEncoded, encoded, `encodeFrame roundtrip failed for ${geometry.type}`);
    }
});

test('GeometryFrame.toGeometry() recovers original geometry (web)', async () => {
    const wkp = await createWkp();

    for (const geometry of GEOMETRIES) {
        const encoded = wkp.encode(geometry, 6);
        const frame = wkp.decodeFrame(encoded);
        const recovered = frame.toGeometry();
        assertGeometryClose(recovered, geometry);
    }
});

test('GeometryFrame.toBuffer() / fromBuffer() roundtrip (web)', async () => {
    const wkp = await createWkp();
    const geometry = {
        type: 'Polygon',
        coordinates: [[[0, 0], [10, 0], [10, 10], [0, 10], [0, 0]], [[2, 2], [4, 2], [4, 4], [2, 4], [2, 2]]]
    };
    const encoded = wkp.encode(geometry, 6);
    const frame = wkp.decodeFrame(encoded);

    const buf = frame.toBuffer();
    assert.ok(buf instanceof ArrayBuffer, 'toBuffer() should return ArrayBuffer');

    const recovered = wkp.GeometryFrame.fromBuffer(buf);
    assert.equal(recovered.version, frame.version);
    assert.equal(recovered.precision, frame.precision);
    assert.equal(recovered.dimensions, frame.dimensions);
    assert.equal(recovered.geometryType, frame.geometryType);
    assert.equal(recovered.coords.length, frame.coords.length);
    for (let i = 0; i < frame.coords.length; i += 1) {
        assertClose(recovered.coords[i], frame.coords[i]);
    }

    // Verify geometry survives buffer roundtrip
    const geomRecovered = recovered.toGeometry();
    assertGeometryClose(geomRecovered, geometry);
});

test('GeometryFrame buffer is 40-byte header + data (web)', async () => {
    const wkp = await createWkp();
    const geometry = { type: 'Point', coordinates: [1.0, 2.0] };
    const encoded = wkp.encode(geometry, 6);
    const frame = wkp.decodeFrame(encoded);
    const buf = frame.toBuffer();

    const expectedSize = 40
        + frame.coords.byteLength
        + frame.segmentPointCounts.byteLength
        + frame.groupSegmentCounts.byteLength;
    assert.equal(buf.byteLength, expectedSize);
});

test('cross-language buffer format matches Python and Node bindings (web)', async () => {
    // Geometry: LineString [(0,0),(1,1)] at precision=6.
    // Coordinates 0.0 and 1.0 are exact IEEE-754 float64 values and survive
    // WKP encode/decode unchanged, making the expected bytes deterministic.
    const wkp = await createWkp();
    const geometry = { type: 'LineString', coordinates: [[0, 0], [1, 1]] };
    const encoded = wkp.encode(geometry, 6);
    const frame = wkp.decodeFrame(encoded);
    const buf = frame.toBuffer();

    // prettier-ignore
    const expected = new Uint8Array([
        0x01, 0x00, 0x00, 0x00,                          // version=1        (int32 LE)
        0x06, 0x00, 0x00, 0x00,                          // precision=6      (int32 LE)
        0x02, 0x00, 0x00, 0x00,                          // dimensions=2     (int32 LE)
        0x02, 0x00, 0x00, 0x00,                          // geometry_type=2  (int32 LE, LINESTRING)
        0x04, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // coord_value_count=4 (uint64 LE)
        0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // segment_count=1  (uint64 LE)
        0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // group_count=1    (uint64 LE)
        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // coords[0][0]=0.0 (float64 LE)
        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // coords[0][1]=0.0 (float64 LE)
        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xF0, 0x3F, // coords[1][0]=1.0 (float64 LE)
        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xF0, 0x3F, // coords[1][1]=1.0 (float64 LE)
        0x02, 0x00, 0x00, 0x00,                          // segment_point_counts=[2] (uint32 LE)
        0x01, 0x00, 0x00, 0x00,                          // group_segment_counts=[1] (uint32 LE)
    ]);

    assert.equal(buf.byteLength, 80);
    const actual = new Uint8Array(buf);
    for (let i = 0; i < expected.length; i += 1) {
        assert.equal(actual[i], expected[i], `byte ${i} mismatch — cross-language buffer format has changed`);
    }

    // Verify fromBuffer recovers the geometry from these canonical bytes
    const recovered = wkp.GeometryFrame.fromBuffer(expected.buffer);
    assert.equal(recovered.version, 1);
    assert.equal(recovered.precision, 6);
    assert.equal(recovered.dimensions, 2);
    assertGeometryClose(recovered.toGeometry(), geometry, 1e-9);
});
